crates/metal-llvm/src/codegen.rs

```diff
@@ -0,0 +1,195 @@
+        match ty {
+            metal_ast::Ty::Primitive(primitive) => self.primitive(primitive),
+            _ => {
+                todo!()
+            }
+        }
+    }
+
+    pub fn function_definition(&self, ty: metal_ast::FnDefStmt, library: bool) -> LLVMValueRef {
+        let ret_ty = ty
+            .return_type
+            .unwrap_or(metal_ast::Ty::Primitive(metal_ast::Primitives::Void));
+        let return_type = self.ty(&ret_ty);
+
+        let mut params = Vec::with_capacity(ty.inputs.len());
+
+        for fn_input in ty.inputs {
+            params.push(self.ty(&fn_input.ty))
+        }
+
+        let linkage = if ty.public {
+            LLVMLinkage::LLVMExternalLinkage
+        } else {
+            LLVMLinkage::LLVMInternalLinkage
+        };
+
+        let fun_name: String = if library {
+            let mut name = "".to_string();
+            for module in self.module_name.split('.') {
+                let m1 = module.len().to_string() + module;
+                let m = m1.as_str();
+                name += m;
+            }
+            let n = name + ty.ident.inner.len().to_string().as_str() + ty.ident.inner;
+            n
+        } else {
+            // this only applies to
+            // main.mt, the entry point of a program.
+            ty.ident.inner.to_string()
+        };
+
+        unsafe {
+            // TODO: variadic argument support
+            // like `*args`
+            let c_fun_name = CString::new(fun_name).unwrap();
+
+            let function = LLVMAddFunction(
+                self.module,
+                c_fun_name.as_ptr(),
+                LLVMFunctionType(return_type, params.as_mut_ptr(), params.len() as c_uint, 0),
+            );
+            LLVMSetLinkage(function, linkage);
+
+            let c_entry = CString::new("entry").unwrap();
+
+            let entry_block = LLVMAppendBasicBlockInContext(self.ctx, function, c_entry.as_ptr());
+            LLVMPositionBuilderAtEnd(self.builder, entry_block);
+
+            // TODO: global variables
+            let variables = BTreeMap::new();
+
+            for stmt in ty.body {
+                match stmt {
+                    // These aren't allowed inside of function bodies
+                    metal_ast::Statement::ClassDef(_) => todo!(),
+                    metal_ast::Statement::FnDef(_) => todo!(),
+                    metal_ast::Statement::Import(_) => panic!(),
+
+                    metal_ast::Statement::Expr(expr) => {
+                        self.expression(expr.expr, &variables);
+                    }
+                }
+            }
+
+            function
+        }
+    }
+
+    pub fn expression(
+        &self,
+        expr: metal_ast::Expr,
+        variables: &BTreeMap<String, Variable>,
+    ) -> LLVMValueRef {
+        match expr {
+            metal_ast::Expr::Number(num) => unsafe {
+                let (sign_extend, value) = if num.value.is_negative() {
+                    (1, -num.value as u64)
```

Comment:
I'm not sure you need to do negation plus cast here. LLVM accepts a "pure" u64, and judging by it's code, it will do it's own thing without our interference. But again, not entirely sure

Reply:
this is mostly turning the AST's integer value to a positive number so it doesn't error out when converted for LLVM.

Reply:
Oh yeah sorry you're right, not sure what got me